### ISSUE TITLE
Should we run sbml test suite in parallel? 

### ISSUE THREAD

**Issue by CiaranWelsh** [sys-bio/roadrunner#848]
With C++17 we can get some degree of parallelism just by using the [parallel execution policy](https://en.cppreference.com/w/cpp/algorithm/execution_policy_tag_t) for [for_each](https://en.cppreference.com/w/cpp/algorithm/for_each). We can process the sbml test suite a bit quicker with something like the following: 

```

TEST_F(SbmlTestSuite, RunTests){
    // list of all disabled tests
    std::vector<int> exclusion({
        533, 534, 536, 537, 538, 569, 570, 575, 663, 664, 762, 777, 844, 983,
        993, 1044, 1108, 1165, 1167, 1168, 1198, 1208, 1234, 1235, 1236, 1238,
        1239, 1241, 1242, 1244, 1248, 1249, 1250, 1251, 1252, 1253, 1254, 1255,
        1256, 1257, 1258, 1259, 1260, 1261, 1262, 1263, 1264, 1265, 1266, 1267,
        1268, 1269, 1270, 1282, 1283, 1284, 1285, 1286, 1288, 1289, 1290, 1291,
        1292, 1293, 1294, 1295, 1296, 1297, 1298, 1299, 1321, 1322, 1350, 1359,
        1368, 1377, 1386, 1398, 1400, 1401, 1402, 1403, 1405, 1406, 1408, 1409,
        1416, 1419, 1444, 1445, 1446, 1447, 1448, 1449, 1450, 1451, 1452, 1453,
        1454, 1455, 1456, 1457, 1458, 1459, 1460, 1461, 1462, 1463, 1464, 1465,
        1471, 1472, 1473, 1475, 1476, 1477, 1478, 1479, 1482, 1483, 1484, 1488,
        1498, 1499, 1500, 1501, 1502, 1503, 1504, 1505, 1506, 1507, 1508, 1509,
        1510, 1511, 1512, 1513, 1514, 1517, 1525, 1526, 1527, 1528, 1529, 1540,
        1541, 1542, 1543, 1552, 1553, 1554, 1556, 1567, 1571, 1574, 1575, 1576,
        1577, 1579, 1589, 1626, 1627, 1628, 1629, 1630, 1635, 1636, 1657, 1727,
        1728, 1729, 1736, 1737, 1738, 1778, 1780});
    int lastTest = 1809;
    std::vector<int> cases;

    // exclude disabled tests
    for (int i=1; i<=lastTest; i++) {
        if (std::find(exclusion.begin(), exclusion.end(), i) == exclusion.end()){
            cases.push_back(i);
        }
    }
    // execute in parallel
    std::for_each(std::execution::par, cases.begin(), cases.end(), [&](int i){
        EXPECT_TRUE(RunTest(i));
    });
}

```

The advantage is quicker build times because the sbml test suite is be far the longest set of tests we run. The disadvantage is it may be slightly harder to locate bugs when they do occur. We would need to use the single test one at a time with failed tests.

[Here's](https://github.com/sys-bio/roadrunner/tree/parallel-sbml-test-suite) the branch.

**Comment by luciansmith** [comment]
That works for me.  I always set the single test to the test I'm debugging anyway.

The only enhancement I could see would be to go ahead and explicitly list and disable all the skipped tests, so that the report mentions the disabled tests so we don't forget, and it's a little easier to un-disable-and-test them going forward?

**Comment by CiaranWelsh** [comment]
Sounds good but 

>explicitly list and disable all the skipped tests

How do we do this? Do you mean just print out "case x is disabled" for all the disabled tests?

**Comment by luciansmith** [comment]
Just add in:

TEST_F(SbmlTestSuite, DISABLED_t1465)
{
    // Unknown throw.
    EXPECT_TRUE(RunTest(1465));
}

for the various disabled test numbers.

(Note that tests that we know we don't pass are run normally, i.e. if it has delay in it.  All the tests we're skipping are bugs, because we claim we should be able to pass them.)

**Comment by CiaranWelsh** [comment]
Oh I see, yep we can do that.

**Comment by CiaranWelsh** [comment]
Okay so it turns out that the clang compiler doesn't yet support parallel execution policies in c++17 or perhaps even c++20 by the [looks of it](https://en.cppreference.com/w/cpp/compiler_support) (search for parallel). Given this, and that the sbml test suite tests are easily parallelisable (i.e. embarrassingly parallel) I think its worth figuring out how to use  a queue along with the thread library instead. However, this is not a priority for right now so I'll put this to one side.

**Comment by luciansmith** [comment]
This got slightly easier to do because now there are no longer any disabled tests in our runs of the SBML Test Suite; we just check the tags of the test and do a full test or a spot test accordingly.

**Comment by CiaranWelsh** [comment]
Since the pickle work is basically done, I think I'll return to this towards the end of the week. Should be some good practice with the threading library!

**Comment by luciansmith** [comment]
The branch in question was deleted; I think in the end it was too awkward.  Since we mostly have to run this on azure where we don't have access to multiple cores anyway, I'm closing this.
